framework/src/main/java/org/checkerframework/common/basetype/BaseTypeChecker.java

```diff
@@ -497,14 +505,99 @@ public void typeProcess(TypeElement element, TreePath tree) {
 
         this.errsOnLastExit = nerrorsOfAllPreviousCheckers;
         super.typeProcess(element, tree);
-
+        if (this.parentChecker == null) {
+            warnUnneededSuppressions();
+        }
         if (getSubcheckers().size() > 0) {
             printCollectedMessages(tree.getCompilationUnit());
             // Update errsOnLastExit to reflect the errors issued.
             this.errsOnLastExit = log.nerrors;
         }
     }
 
+    /**
+     * Issues a warning about any {@code @SuppressWarnings} that isn't used by this checker, but
+     * contains a key that would suppress a warning from this checker.
+     */
+    private void warnUnneededSuppressions() {
+        if (!hasOption("warnUnneededSuppressions")) {
+            return;
+        }
+        Set<Element> elementsSuppress = new HashSet<>(this.elementsWithSuppressedWarnings);
+        this.elementsWithSuppressedWarnings.clear();
+        Set<String> checkerKeys = new HashSet<>(getSuppressWarningsKeys());
+        Set<String> errorKeys = new HashSet<>(messages.stringPropertyNames());
+        for (BaseTypeChecker subChecker : subcheckers) {
+            elementsSuppress.addAll(subChecker.elementsWithSuppressedWarnings);
+            subChecker.elementsWithSuppressedWarnings.clear();
+            checkerKeys.addAll(subChecker.getSuppressWarningsKeys());
+            errorKeys.addAll(subChecker.messages.stringPropertyNames());
+            subChecker.getVisitor().treesWithSuppressWarnings.clear();
+        }
+        // It's not clear for which checker this suppression is intended,
+        // so never report it as unused.
+        checkerKeys.remove(SourceChecker.SUPPRESS_ALL_KEY);
+
+        for (Tree tree : getVisitor().treesWithSuppressWarnings) {
```

Comment:
I can write `@SuppressWarnings("purity")` to suppress all purity warnings.
However, `"purity"` isn't a message key. Would this logic notice if it is unnecessary?

Reply:
Good point.  It would not have, but I fixed it.  (I also add a test case.)